src/main/scala/io/qbeast/spark/index/OTreeDataAnalyzer.scala

```diff
@@ -46,25 +48,30 @@ object DoublePassOTreeDataAnalyzer extends OTreeDataAnalyzer with Serializable {
   private lazy val logger = org.apache.log4j.LogManager.getLogger(this.getClass)
 
   private[index] def calculateRevisionChanges(
-      data: DataFrame,
+      dataFrameStats: Array[Row],
       revision: Revision): Option[RevisionChange] = {
 
-    val columnStats = revision.columnTransformers.map(_.stats)
-    val columnsExpr = columnStats.flatMap(_.columns)
-    def needStats = columnsExpr.nonEmpty
-
-    val newTransformation = if (!needStats) {
-      revision.columnTransformers.map(_.makeTransformation(identity))
-    } else {
-      // This is a actions that will be executed on the dataframe
-      val rows = data.selectExpr(columnsExpr: _*).collect()
-      if (rows.isEmpty) {
-        throw new RuntimeException(
-          "The DataFrame is empty, why are you trying to index an empty dataset?")
+    val columnNames = revision.columnTransformers.map(_.columnName)
+
+    val mins = dataFrameStats.filter(_.get(0) == "min").head.getValuesMap(columnNames)
+    val maxs = dataFrameStats.filter(_.get(0) == "max").head.getValuesMap(columnNames)
+
+    val newTransformation = revision.columnTransformers.map(ct => {
+      val minVal: String = mins(ct.columnName)
+      val maxVal: String = maxs(ct.columnName)
+      val columnStats: ColumnStats = {
```

Comment:
I think it would be useful to build these ColumnStats from the companion object. Also, instead of initializing min and max values outside the map, we can retrieve all the stats for the column at the same time

It would be something like:

```scala
val newTransformation = revision.columnTransformers.map(ct => {
      val stats =
        dataFrameStats.map(row => (row.getAs[String](0), row.getAs[String](ct.columnName))).toMap // a map to get the stats by their name
      val columnStats = ColumnStats(stats, ct.dataType) // here in the apply method we can do all this matching with the data types
      ct.makeTransformation(columnStats)
    })
```

Reply:
Having the `stats` in `Map[String, String] ` can help to get the information we want through: 

```scala
val min = stats("min")
val max = stats("max")
```

And we can as well collect the other properties like `stddev`, `mean` and `count`

Reply:
Solved in b9cff7b30c0f53bcf067c451d9455a98fe1ab68e